File: viddhana/packages/sdks/python/src/viddhana/governance.py
"""Governance module for proposals, voting, and delegation."""

from typing import Optional, Dict, Any, List, TYPE_CHECKING
from web3 import Web3

from .types import Proposal, Vote, Delegation, TransactionReceipt

if TYPE_CHECKING:
    from eth_account.signers.local import LocalAccount


# Governance contract ABI (simplified)
GOVERNOR_ABI = [
    {
        "name": "propose",
        "type": "function",
        "inputs": [
            {"name": "targets", "type": "address[]"},
            {"name": "values", "type": "uint256[]"},
            {"name": "calldatas", "type": "bytes[]"},
            {"name": "description", "type": "string"},
        ],
        "outputs": [{"name": "proposalId", "type": "uint256"}],
    },
    {
        "name": "castVote",
        "type": "function",
        "inputs": [
            {"name": "proposalId", "type": "uint256"},
            {"name": "support", "type": "uint8"},
        ],
        "outputs": [{"name": "", "type": "uint256"}],
    },
    {
        "name": "castVoteWithReason",
        "type": "function",
        "inputs": [
            {"name": "proposalId", "type": "uint256"},
            {"name": "support", "type": "uint8"},
            {"name": "reason", "type": "string"},
        ],
        "outputs": [{"name": "", "type": "uint256"}],
    },
    {
        "name": "delegate",
        "type": "function",
        "inputs": [{"name": "delegatee", "type": "address"}],
        "outputs": [],
    },
    {
        "name": "getVotes",
        "type": "function",
        "inputs": [{"name": "account", "type": "address"}],
        "outputs": [{"name": "", "type": "uint256"}],
    },
    {
        "name": "proposals",
        "type": "function",
        "inputs": [{"name": "proposalId", "type": "uint256"}],
        "outputs": [
            {"name": "id", "type": "uint256"},
            {"name": "proposer", "type": "address"},
            {"name": "eta", "type": "uint256"},
            {"name": "startBlock", "type": "uint256"},
            {"name": "endBlock", "type": "uint256"},
            {"name": "forVotes", "type": "uint256"},
            {"name": "againstVotes", "type": "uint256"},
            {"name": "abstainVotes", "type": "uint256"},
            {"name": "canceled", "type": "bool"},
            {"name": "executed", "type": "bool"},
        ],
    },
]


class GovernanceModule:
    """Module for governance operations."""

    # Contract addresses (configurable per network)
    GOVERNOR_ADDRESS = "0x0000000000000000000000000000000000001003"
    VDH_TOKEN_ADDRESS = "0x0000000000000000000000000000000000001000"

    def __init__(self, w3: Web3, account: Optional["LocalAccount"] = None):
        """
        Initialize GovernanceModule.

        Args:
            w3: Web3 instance connected to Atlas chain.
            account: Optional account for signing transactions.
        """
        self._w3 = w3
        self._account = account

    def _get_governor(self):
        """Get Governor contract instance."""
        return self._w3.eth.contract(
            address=Web3.to_checksum_address(self.GOVERNOR_ADDRESS),
            abi=GOVERNOR_ABI,
        )

    def _sign_and_send(self, tx: Dict[str, Any]) -> TransactionReceipt:
        """Sign and send a transaction."""
        if not self._account:
            raise ValueError("Account required for transactions")

        tx["from"] = self._account.address
        tx["nonce"] = self._w3.eth.get_transaction_count(self._account.address)
        tx["gas"] = self._w3.eth.estimate_gas(tx)
        tx["gasPrice"] = self._w3.eth.gas_price

        signed = self._account.sign_transaction(tx)
        tx_hash = self._w3.eth.send_raw_transaction(signed.raw_transaction)
        receipt = self._w3.eth.wait_for_transaction_receipt(tx_hash)

        return TransactionReceipt(
            tx_hash=receipt["transactionHash"].hex(),
            block_number=receipt["blockNumber"],
            status=receipt["status"] == 1,
            gas_used=receipt["gasUsed"],
            effective_gas_price=receipt.get("effectiveGasPrice", 0),
            logs=[dict(log) for log in receipt["logs"]],
        )

    def create_proposal(
        self,
        targets: List[str],
        values: List[int],
        calldatas: List[bytes],
        description: str,
    ) -> TransactionReceipt:
        """
        Create a new governance proposal.

        Args:
            targets: List of target contract addresses.
            values: List of ETH values to send with each call.
            calldatas: List of encoded function calls.
            description: Proposal description (markdown supported).

        Returns:
            Transaction receipt containing the proposal ID in logs.

        Raises:
            ValueError: If no account is configured.
        """
        if not self._account:
            raise ValueError("Account required to create proposal")

        governor = self._get_governor()
        checksum_targets = [Web3.to_checksum_address(t) for t in targets]

        tx = governor.functions.propose(
            checksum_targets,
            values,
            calldatas,
            description,
        ).build_transaction({
            "chainId": self._w3.eth.chain_id,
        })

        return self._sign_and_send(tx)

    def cast_vote(
        self,
        proposal_id: int,
        support: int,
        reason: Optional[str] = None,
    ) -> TransactionReceipt:
        """
        Cast a vote on a proposal.

        Args:
            proposal_id: The proposal ID to vote on.
            support: Vote type (0=against, 1=for, 2=abstain).
            reason: Optional reason for the vote.

        Returns:
            Transaction receipt.

        Raises:
            ValueError: If no account is configured or invalid support value.
        """
        if not self._account:
            raise ValueError("Account required to vote")

        if support not in (0, 1, 2):
            raise ValueError("Support must be 0 (against), 1 (for), or 2 (abstain)")

        governor = self._get_governor()

        if reason:
            tx = governor.functions.castVoteWithReason(
                proposal_id, support, reason
            ).build_transaction({
                "chainId": self._w3.eth.chain_id,
            })
        else:
            tx = governor.functions.castVote(
                proposal_id, support
            ).build_transaction({
                "chainId": self._w3.eth.chain_id,
            })

        return self._sign_and_send(tx)

    def vote_for(self, proposal_id: int, reason: Optional[str] = None) -> TransactionReceipt:
        """
        Vote in favor of a proposal.

        Args:
            proposal_id: The proposal ID.
            reason: Optional reason for the vote.

        Returns:
            Transaction receipt.
        """
        return self.cast_vote(proposal_id, 1, reason)

    def vote_against(self, proposal_id: int, reason: Optional[str] = None) -> TransactionReceipt:
        """
        Vote against a proposal.

        Args:
            proposal_id: The proposal ID.
            reason: Optional reason for the vote.

        Returns:
            Transaction receipt.
        """
        return self.cast_vote(proposal_id, 0, reason)

    def vote_abstain(self, proposal_id: int, reason: Optional[str] = None) -> TransactionReceipt:
        """
        Abstain from voting on a proposal.

        Args:
            proposal_id: The proposal ID.
            reason: Optional reason.

        Returns:
            Transaction receipt.
        """
        return self.cast_vote(proposal_id, 2, reason)

    def delegate(self, delegatee: str) -> TransactionReceipt:
        """
        Delegate voting power to another address.

        Args:
            delegatee: Address to delegate voting power to.

        Returns:
            Transaction receipt.

        Raises:
            ValueError: If no account is configured.
        """
        if not self._account:
            raise ValueError("Account required to delegate")

        governor = self._get_governor()
        delegatee_address = Web3.to_checksum_address(delegatee)

        tx = governor.functions.delegate(delegatee_address).build_transaction({
            "chainId": self._w3.eth.chain_id,
        })

        return self._sign_and_send(tx)

    def self_delegate(self) -> TransactionReceipt:
        """
        Delegate voting power to self (activate voting).

        Returns:
            Transaction receipt.
        """
        if not self._account:
            raise ValueError("Account required")
        return self.delegate(self._account.address)

    def get_voting_power(self, address: Optional[str] = None) -> int:
        """
        Get voting power for an address.

        Args:
            address: Address to query (defaults to connected account).

        Returns:
            Voting power (in token units).
        """
        if address is None:
            if not self._account:
                raise ValueError("Address required or account must be configured")
            address = self._account.address

        governor = self._get_governor()
        checksum_address = Web3.to_checksum_address(address)
        return governor.functions.getVotes(checksum_address).call()

    def get_proposal(self, proposal_id: int) -> Proposal:
        """
        Get proposal details.

        Args:
            proposal_id: The proposal ID.

        Returns:
            Proposal details.
        """
        result = self._w3.provider.make_request(
            "vdh_getProposal", [proposal_id]
        )

        if "error" in result:
            raise ValueError(result["error"]["message"])

        data = result["result"]

        return Proposal(
            proposal_id=str(data["proposalId"]),
            proposer=data["proposer"],
            title=data.get("title", ""),
            description=data.get("description", ""),
            start_block=data["startBlock"],
            end_block=data["endBlock"],
            for_votes=str(data["forVotes"]),
            against_votes=str(data["againstVotes"]),
            status=data["status"],
        )

    def get_proposals(
        self,
        status: Optional[str] = None,
        limit: int = 10,
        offset: int = 0,
    ) -> List[Proposal]:
        """
        Get list of proposals.

        Args:
            status: Filter by status ('active', 'pending', 'succeeded', 'defeated', 'executed').
            limit: Maximum number of proposals to return.
            offset: Number of proposals to skip.

        Returns:
            List of proposals.
        """
        params: Dict[str, Any] = {"limit": limit, "offset": offset}
        if status:
            params["status"] = status

        result = self._w3.provider.make_request("vdh_getProposals", [params])

        if "error" in result:
            raise ValueError(result["error"]["message"])

        proposals = []
        for data in result["result"]:
            proposals.append(Proposal(
                proposal_id=str(data["proposalId"]),
                proposer=data["proposer"],
                title=data.get("title", ""),
                description=data.get("description", ""),
                start_block=data["startBlock"],
                end_block=data["endBlock"],
                for_votes=str(data["forVotes"]),
                against_votes=str(data["againstVotes"]),
                status=data["status"],
            ))

        return proposals

File: viddhana/packages/sdks/javascript/src/modules/governance.ts
/**
 * Governance Module - Proposals, voting, and delegation
 */

import { ethers } from 'ethers';
import type {
  Proposal,
  ProposalState,
  VoteReceipt,
  DelegationInfo,
} from '../types';

// Contract ABIs
const GOVERNOR_ABI = [
  'function propose(address[] targets, uint256[] values, bytes[] calldatas, string description) returns (uint256)',
  'function castVote(uint256 proposalId, uint8 support) returns (uint256)',
  'function castVoteWithReason(uint256 proposalId, uint8 support, string reason) returns (uint256)',
  'function queue(uint256 proposalId) returns (uint256)',
  'function execute(uint256 proposalId) returns (uint256)',
  'function cancel(uint256 proposalId) returns (uint256)',
  'function state(uint256 proposalId) view returns (uint8)',
  'function proposalVotes(uint256 proposalId) view returns (uint256 againstVotes, uint256 forVotes, uint256 abstainVotes)',
  'function proposalSnapshot(uint256 proposalId) view returns (uint256)',
  'function proposalDeadline(uint256 proposalId) view returns (uint256)',
  'function getVotes(address account, uint256 blockNumber) view returns (uint256)',
  'function hasVoted(uint256 proposalId, address account) view returns (bool)',
  'function proposals(uint256 proposalId) view returns (tuple(address proposer, uint256 eta, uint256 startBlock, uint256 endBlock, uint256 forVotes, uint256 againstVotes, uint256 abstainVotes, bool canceled, bool executed))',
];

const VDH_TOKEN_ABI = [
  'function delegate(address delegatee) returns ()',
  'function delegates(address account) view returns (address)',
  'function getVotes(address account) view returns (uint256)',
  'function getPastVotes(address account, uint256 blockNumber) view returns (uint256)',
  'function getPastTotalSupply(uint256 blockNumber) view returns (uint256)',
];

// Contract addresses
const CONTRACT_ADDRESSES = {
  mainnet: {
    governor: '0x0000000000000000000000000000000000000003',
    vdhToken: '0x0000000000000000000000000000000000000004',
  },
  testnet: {
    governor: '0x0000000000000000000000000000000000000003',
    vdhToken: '0x0000000000000000000000000000000000000004',
  },
};

const PROPOSAL_STATES: ProposalState[] = [
  'Pending',
  'Active',
  'Canceled',
  'Defeated',
  'Succeeded',
  'Queued',
  'Expired',
  'Executed',
];

/**
 * Governance module for proposals, voting, and delegation
 *
 * @example
 * ```typescript
 * // Create a proposal
 * const tx = await client.governance.createProposal(
 *   ['0xTargetContract...'],
 *   [0n],
 *   ['0x...'],
 *   'Proposal to upgrade the protocol'
 * );
 *
 * // Cast a vote
 * await client.governance.castVote(proposalId, 1); // 1 = For
 *
 * // Delegate voting power
 * await client.governance.delegate('0xDelegate...');
 * ```
 */
export class Governance {
  private readonly provider: ethers.JsonRpcProvider;
  private readonly signer: ethers.Wallet | null;

  constructor(
    provider: ethers.JsonRpcProvider,
    signer: ethers.Wallet | null
  ) {
    this.provider = provider;
    this.signer = signer;
  }

  /**
   * Creates a new governance proposal
   *
   * @param targets - Array of target contract addresses
   * @param values - Array of ETH values to send
   * @param calldatas - Array of encoded function calls
   * @param description - Proposal description
   * @returns Transaction response
   */
  async createProposal(
    targets: string[],
    values: bigint[],
    calldatas: string[],
    description: string
  ): Promise<ethers.TransactionResponse> {
    this.requireSigner();

    const governor = new ethers.Contract(
      CONTRACT_ADDRESSES.mainnet.governor,
      GOVERNOR_ABI,
      this.signer!
    );

    const tx = await governor.propose(targets, values, calldatas, description);
    return tx as ethers.TransactionResponse;
  }

  /**
   * Casts a vote on a proposal
   *
   * @param proposalId - The proposal ID
   * @param support - Vote type: 0 = Against, 1 = For, 2 = Abstain
   * @returns Transaction response
   */
  async castVote(
    proposalId: string | bigint,
    support: 0 | 1 | 2
  ): Promise<ethers.TransactionResponse> {
    this.requireSigner();

    const governor = new ethers.Contract(
      CONTRACT_ADDRESSES.mainnet.governor,
      GOVERNOR_ABI,
      this.signer!
    );

    const tx = await governor.castVote(proposalId, support);
    return tx as ethers.TransactionResponse;
  }

  /**
   * Casts a vote with a reason
   *
   * @param proposalId - The proposal ID
   * @param support - Vote type: 0 = Against, 1 = For, 2 = Abstain
   * @param reason - Reason for the vote
   * @returns Transaction response
   */
  async castVoteWithReason(
    proposalId: string | bigint,
    support: 0 | 1 | 2,
    reason: string
  ): Promise<ethers.TransactionResponse> {
    this.requireSigner();

    const governor = new ethers.Contract(
      CONTRACT_ADDRESSES.mainnet.governor,
      GOVERNOR_ABI,
      this.signer!
    );

    const tx = await governor.castVoteWithReason(proposalId, support, reason);
    return tx as ethers.TransactionResponse;
  }

  /**
   * Delegates voting power to another address
   *
   * @param delegatee - Address to delegate to
   * @returns Transaction response
   */
  async delegate(delegatee: string): Promise<ethers.TransactionResponse> {
    this.requireSigner();

    const vdhToken = new ethers.Contract(
      CONTRACT_ADDRESSES.mainnet.vdhToken,
      VDH_TOKEN_ABI,
      this.signer!
    );

    const tx = await vdhToken.delegate(delegatee);
    return tx as ethers.TransactionResponse;
  }

  /**
   * Gets proposal details
   *
   * @param proposalId - The proposal ID
   * @returns Proposal information
   */
  async getProposal(proposalId: string | bigint): Promise<Proposal> {
    const governor = new ethers.Contract(
      CONTRACT_ADDRESSES.mainnet.governor,
      GOVERNOR_ABI,
      this.provider
    );

    const [proposalData, state, votes] = await Promise.all([
      governor.proposals(proposalId),
      governor.state(proposalId),
      governor.proposalVotes(proposalId),
    ]);

    return {
      id: proposalId.toString(),
      proposer: proposalData.proposer as string,
      targets: [], // Would need separate storage/events to retrieve
      values: [],
      calldatas: [],
      description: '', // Would need to retrieve from events
      state: PROPOSAL_STATES[state as number] ?? 'Pending',
      startBlock: Number(proposalData.startBlock),
      endBlock: Number(proposalData.endBlock),
      forVotes: votes.forVotes as bigint,
      againstVotes: votes.againstVotes as bigint,
      abstainVotes: votes.abstainVotes as bigint,
    };
  }

  /**
   * Gets the state of a proposal
   *
   * @param proposalId - The proposal ID
   * @returns Proposal state
   */
  async getProposalState(proposalId: string | bigint): Promise<ProposalState> {
    const governor = new ethers.Contract(
      CONTRACT_ADDRESSES.mainnet.governor,
      GOVERNOR_ABI,
      this.provider
    );

    const state = await governor.state(proposalId);
    return PROPOSAL_STATES[state as number] ?? 'Pending';
  }

  /**
   * Gets vote receipt for an account on a proposal
   *
   * @param proposalId - The proposal ID
   * @param account - The voter's address
   * @returns Vote receipt
   */
  async getVoteReceipt(
    proposalId: string | bigint,
    account: string
  ): Promise<VoteReceipt> {
    const governor = new ethers.Contract(
      CONTRACT_ADDRESSES.mainnet.governor,
      GOVERNOR_ABI,
      this.provider
    );

    const hasVoted = await governor.hasVoted(proposalId, account);
    const snapshot = await governor.proposalSnapshot(proposalId);
    const votes = await governor.getVotes(account, snapshot);

    return {
      hasVoted: hasVoted as boolean,
      support: 0, // Would need to retrieve from events
      votes: votes as bigint,
    };
  }

  /**
   * Gets delegation info for an account
   *
   * @param account - The account address
   * @returns Delegation information
   */
  async getDelegation(account: string): Promise<DelegationInfo> {
    const vdhToken = new ethers.Contract(
      CONTRACT_ADDRESSES.mainnet.vdhToken,
      VDH_TOKEN_ABI,
      this.provider
    );

    const [delegatee, votingPower] = await Promise.all([
      vdhToken.delegates(account),
      vdhToken.getVotes(account),
    ]);

    return {
      delegator: account,
      delegatee: delegatee as string,
      votingPower: votingPower as bigint,
      delegatedAt: 0, // Would need to retrieve from events
    };
  }

  /**
   * Gets current voting power for an account
   *
   * @param account - The account address
   * @returns Current voting power
   */
  async getVotingPower(account: string): Promise<bigint> {
    const vdhToken = new ethers.Contract(
      CONTRACT_ADDRESSES.mainnet.vdhToken,
      VDH_TOKEN_ABI,
      this.provider
    );

    return vdhToken.getVotes(account) as Promise<bigint>;
  }

  /**
   * Gets past voting power at a specific block
   *
   * @param account - The account address
   * @param blockNumber - Block number to check
   * @returns Voting power at that block
   */
  async getPastVotingPower(
    account: string,
    blockNumber: number
  ): Promise<bigint> {
    const vdhToken = new ethers.Contract(
      CONTRACT_ADDRESSES.mainnet.vdhToken,
      VDH_TOKEN_ABI,
      this.provider
    );

    return vdhToken.getPastVotes(account, blockNumber) as Promise<bigint>;
  }

  /**
   * Queues a successful proposal for execution
   *
   * @param proposalId - The proposal ID
   * @returns Transaction response
   */
  async queueProposal(
    proposalId: string | bigint
  ): Promise<ethers.TransactionResponse> {
    this.requireSigner();

    const governor = new ethers.Contract(
      CONTRACT_ADDRESSES.mainnet.governor,
      GOVERNOR_ABI,
      this.signer!
    );

    const tx = await governor.queue(proposalId);
    return tx as ethers.TransactionResponse;
  }

  /**
   * Executes a queued proposal
   *
   * @param proposalId - The proposal ID
   * @returns Transaction response
   */
  async executeProposal(
    proposalId: string | bigint
  ): Promise<ethers.TransactionResponse> {
    this.requireSigner();

    const governor = new ethers.Contract(
      CONTRACT_ADDRESSES.mainnet.governor,
      GOVERNOR_ABI,
      this.signer!
    );

    const tx = await governor.execute(proposalId);
    return tx as ethers.TransactionResponse;
  }

  /**
   * Cancels a proposal
   *
   * @param proposalId - The proposal ID
   * @returns Transaction response
   */
  async cancelProposal(
    proposalId: string | bigint
  ): Promise<ethers.TransactionResponse> {
    this.requireSigner();

    const governor = new ethers.Contract(
      CONTRACT_ADDRESSES.mainnet.governor,
      GOVERNOR_ABI,
      this.signer!
    );

    const tx = await governor.cancel(proposalId);
    return tx as ethers.TransactionResponse;
  }

  /**
   * Gets proposal voting period
   *
   * @param proposalId - The proposal ID
   * @returns Start and end blocks
   */
  async getVotingPeriod(
    proposalId: string | bigint
  ): Promise<{ startBlock: number; endBlock: number }> {
    const governor = new ethers.Contract(
      CONTRACT_ADDRESSES.mainnet.governor,
      GOVERNOR_ABI,
      this.provider
    );

    const [startBlock, endBlock] = await Promise.all([
      governor.proposalSnapshot(proposalId),
      governor.proposalDeadline(proposalId),
    ]);

    return {
      startBlock: Number(startBlock),
      endBlock: Number(endBlock),
    };
  }

  /**
   * Checks if an account has voted on a proposal
   *
   * @param proposalId - The proposal ID
   * @param account - The account address
   * @returns True if the account has voted
   */
  async hasVoted(proposalId: string | bigint, account: string): Promise<boolean> {
    const governor = new ethers.Contract(
      CONTRACT_ADDRESSES.mainnet.governor,
      GOVERNOR_ABI,
      this.provider
    );

    return governor.hasVoted(proposalId, account) as Promise<boolean>;
  }

  /**
   * Requires signer to be present
   * @private
   */
  private requireSigner(): void {
    if (!this.signer) {
      throw new Error('Signer required for this operation. Provide privateKey in config.');
    }
  }
}
